aten/src/ATen/native/cpu/batch_norm_kernel.cpp

```diff
@@ -244,64 +244,88 @@ batch_norm_cpu_collect_stats_channels_last_impl(
+        vec::map2<scalar_t>(
+            [](Vec x, Vec y) { return x + y; },
+            buffer_ptr,
+            x_ptr,
+            buffer_ptr,
+            n_channel);
       }
-      scalar_t mean = sum / N;
-      mean_data[c] = mean;
-    }
-  });
+    });
 
-  // compute variance per input, reuse the immediate buffer
-  buffer.zero_();
-  at::parallel_for(0, N, 1, [&](int64_t begin, int64_t end) {
-    int tid = at::get_thread_num();
-    TORCH_CHECK(tid < num_threads, "expect thread id smaller than ", num_threads, ", got thread id ", tid);
-    scalar_t* buffer_ptr = buffer_data + tid * n_channel;
-    for (const auto i : c10::irange(begin, end)) {
-      const scalar_t* x_ptr = input_data + i * n_channel;
-      vec::map3<scalar_t>(
-          [](Vec x, Vec y, Vec mean) { return y + (x - mean) * (x - mean); },
-          buffer_ptr,
-          x_ptr,
-          buffer_ptr,
-          mean_data,
-          n_channel);
-    }
-  });
+    at::parallel_for(0, n_channel, 1, [&](int64_t begin, int64_t end) {
+      for (const auto c : c10::irange(begin, end)) {
+        accscalar_t sum = 0;
+        for (const auto t : c10::irange(num_threads)) {
+          sum += buffer_data[t * n_channel + c];
+        }
+        scalar_t mean = sum / N;
+        mean_data[c] = mean;
+      }
+    });
 
-  at::parallel_for(0, n_channel, 1, [&](int64_t begin, int64_t end) {
-    for (const auto c : c10::irange(begin, end)) {
-      accscalar_t _var_sum = 0;
-      for (const auto t : c10::irange(num_threads)) {
-        _var_sum += buffer_data[t * n_channel + c];
+    // compute variance per input, reuse the immediate buffer
+    buffer.zero_();
+    at::parallel_for(0, N, 1, [&](int64_t begin, int64_t end) {
+      int tid = at::get_thread_num();
+      TORCH_CHECK(tid < num_threads, "expect thread id smaller than ", num_threads, ", got thread id ", tid);
+      scalar_t* buffer_ptr = buffer_data + tid * n_channel;
+      for (const auto i : c10::irange(begin, end)) {
+        const scalar_t* x_ptr = input_data + i * n_channel;
+        vec::map3<scalar_t>(
+            [](Vec x, Vec y, Vec mean) { return y + (x - mean) * (x - mean); },
+            buffer_ptr,
+            x_ptr,
+            buffer_ptr,
+            mean_data,
+            n_channel);
       }
-      var_sum_data[c] = _var_sum;
-    }
-  });
+    });
+
+    at::parallel_for(0, n_channel, 1, [&](int64_t begin, int64_t end) {
+      for (const auto c : c10::irange(begin, end)) {
+        accscalar_t _var_sum = 0;
+        for (const auto t : c10::irange(num_threads)) {
+          _var_sum += buffer_data[t * n_channel + c];
+        }
+        var_sum_data[c] = _var_sum;
+      }
+    });
+  } else {
+    at::parallel_for(0, n_channel, 1, [&](int64_t begin, int64_t end) {
+      for (const auto c : c10::irange(begin, end)) {
+        accscalar_t sum = 0;
+        for (const auto t : c10::irange(N)) {
+          sum += input_data[t * n_channel + c];
```

Comment:
does it make sense to vectorize on c?

Reply:
Hi @jgong5, in channels last format, it is a vertical reduction -- data in each thread is not contiguous. So I think the inner loop is not vectorizeable?

Reply:
> Hi @jgong5, in channels last format, it is a vertical reduction -- data in each thread is not contiguous. So I think the inner loop is not vectorizeable?

I guess we can do tiling on `c` and then do vertical reduction on the tiles which can be vectorizable?

Reply:
Hi @jgong5, we can tile on `c` to vectorize on `c`. But I am unsure of the performance tradeoff as tiling will thin-out the outerloop, reducing the number of iterations in the outerloop from `n_channel` to roughly `n_channel / TILE_SIZE` (i.e., less parallelism). Below is a preliminary benchmark result. 

Tested on 28 physical cores/socket, 1 socket on Skylake.

|                    	| Latency (ms)                                              	|                                                           	|                                            	|                                                                                          	|                                            	|
|--------------------	|-----------------------------------------------------------	|-----------------------------------------------------------	|--------------------------------------------	|------------------------------------------------------------------------------------------	|--------------------------------------------	|
| **Shape (N, C, H, W)** 	| **Baseline (original implementation):   two-path reduction**  	| **[Method 1](https://github.com/pytorch/pytorch/blob/0c9c5d9af7a43edcd0a02d5aa6f865d99b95f1de/aten/src/ATen/native/cpu/batch_norm_kernel.cpp#L308-L330): direct single reduction when N   <= num_threads** 	| **Method 1 Speedup Ratio (Method 1/Baseline)** 	| **[Method 2](https://github.com/pytorch/pytorch/blob/0c9c5d9af7a43edcd0a02d5aa6f865d99b95f1de/aten/src/ATen/native/cpu/batch_norm_kernel.cpp#L332-L378): direct single reduction and   vertical tile on n_channel when N <= num_threads** 	| **Method 2 Speedup Ratio (Method 2/Baseline)** 	|
| (1, 8192, 1, 28)   	| 0.109272003                                               	| 0.038852692                                               	| 2.812469318                                	| 0.029630661                                                                              	| 3.687801738                                	|
| (14, 8192, 1, 2)   	| 0.109317303                                               	| 0.039050579                                               	| 2.799377251                                	| 0.027124882                                                                              	| 4.030148545                                	|
| (1, 2048, 1, 28)   	| 0.077867508                                               	| 0.020139217                                               	| 3.866461466                                	| 0.022242069                                                                              	| 3.500911137                                	|
| (14, 2048, 1, 2)   	| 0.078299046                                               	| 0.019345284                                               	| 4.047448854                                	| 0.022482872                                                                              	| 3.482608696                                	|

Notice that Method 2 doesn't always perform better than Method 1. I am unsure of the performance tradeoff between Method 1 vs. Method 2.

Reply:
| **n_channel** 	| **Baseline (original implementation):   two-path reduction** 	| **[Method 1](https://github.com/pytorch/pytorch/blob/e81694fb2e4f53b072f0a6a86f25c27c31ecbe9d/aten/src/ATen/native/cpu/batch_norm_kernel.cpp#L324-L348): direct single reduction when N   <= num_threads** 	| **Method 1 Speedup Ratio (Method 1/Baseline)** 	| **[Method 2](https://github.com/pytorch/pytorch/blob/e81694fb2e4f53b072f0a6a86f25c27c31ecbe9d/aten/src/ATen/native/cpu/batch_norm_kernel.cpp#L349-L397): direct single reduction and   vertical tile on n_channel when N <= num_threads** 	| **Method 2 Speedup Ratio (Method 2/Baseline)** 	| **Method2/Method1 Ratio** 	|
|---------------	|--------------------------------------------------------------	|---------------------------------------------------------------	|------------------------------------------------	|----------------------------------------------------------------------------------------------	|------------------------------------------------	|---------------------	|
| 1048576       	| 13.67034435                                                  	| 3.748085499                                                   	| 3.647287224                                    	| 3.340418339                                                                                  	| 4.092404893                                    	| 1.122041            	|
| 524288        	| 5.230793953                                                  	| 1.71421051                                                    	| 3.051430336                                    	| 0.975320339                                                                                  	| 5.363154794                                    	| 1.757587            	|
| 262144        	| 2.131233215                                                  	| 0.812349319                                                   	| 2.623542809                                    	| 0.368733406                                                                                  	| 5.779875596                                    	| 2.20308             	|
| 131072        	| 0.990390778                                                  	| 0.417199135                                                   	| 2.373904198                                    	| 0.186152458                                                                                  	| 5.320320705                                    	| 2.241169            	|
| 65536         	| 0.422859192                                                  	| 0.217680931                                                   	| 1.942564237                                    	| 0.129938126                                                                                  	| 3.254311927                                    	| 1.675266            	|
| 32768         	| 0.224406719                                                  	| 0.112738609                                                   	| 1.990504589                                    	| 0.073494911                                                                                  	| 3.053364043                                    	| 1.533965            	|
| 16384         	| 0.143647194                                                  	| 0.0613451                                                     	| 2.341624563                                    	| 0.043981075                                                                                  	| 3.266113731                                    	| 1.394807            	|
| 8192          	| 0.10917902                                                   	| 0.039801598                                                   	| 2.743081347                                    	| 0.031847954                                                                                  	| 3.428132954                                    	| 1.249738            	|
| 4096          	| 0.08869648                                                   	| 0.027146339                                                   	| 3.267345863                                    	| 0.025160313                                                                                  	| 3.525253482                                    	| 1.078935            	|
| **2048**      	| 0.075721741                                                  	| 0.019788742                                                   	| 3.826506024                                    	| 0.022304058                                                                                  	| 3.394975949                                    	| **0.887226**        	|
| **1024**      	| 0.06685257                                                   	| 0.019571781                                                   	| 3.415763187                                    	| 0.019619465                                                                                  	| 3.407461417                                    	| **0.99757**         	|
| **512**       	| 0.051283836                                                  	| 0.018868446                                                   	| 2.717968158                                    	| 0.019276142                                                                                  	| 2.660482375                                    	| **0.97885**         	|
| **256**       	| 0.043172836                                                  	| 0.020422935                                                   	| 2.113938828                                    	| 0.019741058                                                                                  	| 2.186956522                                    	| **1.034541**        	|
| **128**       	| 0.042669773                                                  	| 0.018839836                                                   	| 2.264869653                                    	| 0.019609928                                                                                  	| 2.175927052                                    	| **0.960729**        	|
| **64**        	| 0.038774014                                                  	| 0.015070438                                                   	| 2.572852397                                    	| 0.015821457                                                                                  	| 2.450723327                                    	| **0.952532**        	|
| **32**        	| 0.038630962                                                  	| 0.013892651                                                   	| 2.780676163                                    	| 0.014436245                                                                                  	| 2.675970273                                    	| **0.962345**        	|
| 16            	| 0.027766228                                                  	| 0.012683868                                                   	| 2.189097744                                    	| 0.008983612                                                                                  	| 3.090764331                                    	| 1.41189             	|
| 8             	| 0.019891262                                                  	| 0.011627674                                                   	| 1.710682797                                    	| 0.008108616                                                                                  	| 2.453102029                                    	| 1.43399             	|
| 4             	| 0.018217564                                                  	| 0.01080513                                                    	| 1.686010591                                    	| 0.008733273                                                                                  	| 2.085995086                                    	| 1.237237            	|
| 2             	| 0.017716885                                                  	| 0.01077652                                                    	| 1.644026549                                    	| 0.008654594                                                                                  	| 2.047107438     


Based on offline discussion to find some threshold to tile on C or not,  broader perf data was collected for TILE_SIZE=16, C in range [2^1, 2^20], and (N,H,W) = (1,2,14) for all values of C. Values of (N,H,W)=(1,2,14)  were arbitrarily chosen that satisfies the condition N*H*W <= num_threads (28). 

Above perf data shows that for C<=16 (TILE_SIZE) and C>2048, Method 2 performs better than Method 1.
For C>2048, it could be explained by that C is large enough that the benefit from tiling and vectorization outweigh the synchronization overhead.
For C<=16 (TILE_SIZE), it could be explained by that the problem size is small enough (C <= TILE_SIZE, N*H*W <= num_threads), that it's better to launch single thread with vectorization than C number of threads without vectorization.

Now, for 16 (TILE_SIZE) < C <= 2048 is where Method 2 performs slightly worse than Method 1. 

Use the following threshold,

- C <= TILE_SIZE or C > 2048: use Method 2 
- TILE_SIZE < C <= 2048: use Method 1

2048 is an empirically found threshold.